funnel/views/mixins.py

```diff
@@ -130,3 +133,80 @@ def loader(self, profile, project, section):
+        if draft is not None:
+            db.session.delete(draft)
+        else:
+            raise ValueError(_("There is no draft for the given object."))
+
+    def get_draft_data(self, obj=None):
+        """
+        Returns a tuple of the current draft revision and the formdata needed to initialize forms
+        """
+        draft = self.get_draft(obj)
+        if draft is not None:
+            return draft.revision, draft.formdata
+        else:
+            return None, None
+
+    def autosave_post(self, obj=None):
+        """
+        Handles autosave POST requests
+        """
+        obj = obj if obj is not None else self.obj
+        if 'form.revision' not in request.form:
+            # as form.autosave is true, the form should have `form.revision` field even if it's empty
+            return {'error': _("Form must contain a revision ID.")}, 400
+
+        # CSRF check
+        if forms.Form().validate_on_submit():
+            incoming_data = MultiDict(request.form.items(multi=True))
+            client_revision = incoming_data.pop('form.revision')
+            incoming_data.pop('csrf_token', None)
+
+            # find the last draft
+            draft = self.get_draft(obj)
+
+            if draft is not None:
+                if client_revision is None or (client_revision is not None and str(draft.revision) != client_revision):
+                    # draft exists, but the form did not send a revision ID,
+                    # OR revision ID sent by client does not match the last revision ID
+                    return {'error': _("There has been changes to this draft since you last edited it. Please reload.")}, 400
+                elif client_revision is not None and str(draft.revision) == client_revision:
+                    # revision ID sent my client matches, save updated draft data and update revision ID
+                    existing = draft.formdata
+                    for key in incoming_data.keys():
+                        if existing[key] != incoming_data[key]:
+                            existing[key] = incoming_data[key]
+                    draft.body = {'form': existing}
+                    draft.revision = uuid4()
+            elif draft is None and client_revision:
+                # The form contains a revision ID but no draft exists.
+                # Somebody is making autosave requests with an invalid draft ID.
+                return {'error': _("Invalid revision ID or the existing changes have been submitted already. Please reload.")}, 400
+            else:
+                # no draft exists, create one
+                draft = Draft(
+                    table=Project.__tablename__, table_row_id=obj.uuid,
+                    formdata=incoming_data, revision=uuid4()
+                    )
+            db.session.add(draft)
+            db.session.commit()
+            return {'revision': draft.revision}
+        else:
+            return {'error': _("Invalid CSRF token")}, 400
```

Comment:
I recall we had established a convention for JSON responses earlier:

1. Success: `{"status": "ok", "result": {...}}`
2. Error: `{"status": "error", "error": "error_identifier", "error_description": "English message"}`

The first one I'm ambivalent about, since HTTP status (200 vs 400) is a sufficient indicator of whether the request was successful or not. The second is relevant. The front-end should be able to look up a static error code and invoke custom behaviour. The `error_description` is shown when the front-end does not recognise the error.